sourcehut: update all component; lots of fixes

christoph-heiss opened the pull request:
###### Description of changes

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

This updates all sourcehut components to their latest release; as of 23-07-2023.

Additionally, some things changed and had to be adapted. Further, some bug fixes (notably, should fix #199778, #201424 and #201425 IIRC).

I'm currently only running the metasrht and gitsrht services, thus the other are basically untested (apart from running `nixosTests.sourcehut`). I will do that once I have the spare time, but I thought to open a (draft) PR anyway, such that other people can chime in and maybe even help with testing. It's a rather big thing that definitely needs more than one pair of eyes.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->


- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

christoph-heiss committed:
sourcehut.srht: 0.69.0 -> 0.69.15

Signed-off-by: Christoph Heiss <christoph@c8h4.io>

christoph-heiss committed:
sourcehut.hgsrht: 0.31.3 -> 0.32.4

Signed-off-by: Christoph Heiss <christoph@c8h4.io>

christoph-heiss committed:
sourcehut.hubsrht: 0.14.14 -> 0.17.2

Signed-off-by: Christoph Heiss <christoph@c8h4.io>

christoph-heiss committed:
sourcehut.listssrht: 0.51.11 -> 0.57.8

Signed-off-by: Christoph Heiss <christoph@c8h4.io>

christoph-heiss committed:
sourcehut.mansrht: 0.15.26 -> 0.16.1

Signed-off-by: Christoph Heiss <christoph@c8h4.io>

christoph-heiss committed:
sourcehut.metasrht: 0.61.3 -> 0.64.8

Signed-off-by: Christoph Heiss <christoph@c8h4.io>

christoph-heiss committed:
sourcehut.pagessrht: 0.7.4 -> 0.13.0

Signed-off-by: Christoph Heiss <christoph@c8h4.io>

christoph-heiss committed:
sourcehut: make script work with non-python modules

Signed-off-by: Christoph Heiss <christoph@c8h4.io>

christoph-heiss committed:
sourcehut: default gqlgenVersion 0.17.2 -> 0.17.20

Most packages now uses this version, so avoid having it specify it every
time.

Signed-off-by: Christoph Heiss <christoph@c8h4.io>

christoph-heiss committed:
sourcehut.scmsrht: 0.22.22 -> 0.22.23

Signed-off-by: Christoph Heiss <christoph@c8h4.io>

christoph-heiss committed:
sourcehut.pastesrht: 0.13.8 -> 0.15.1

Signed-off-by: Christoph Heiss <christoph@c8h4.io>

christoph-heiss committed:
sourcehut.todosrht: 0.72.2 -> 0.74.6

Signed-off-by: Christoph Heiss <christoph@c8h4.io>

christoph-heiss committed:
sourcehut: add override for SQLAlchemy 1.x

Signed-off-by: Christoph Heiss <christoph@c8h4.io>

christoph-heiss committed:
sourcehut: make /query endpoint config common to all services

Signed-off-by: Christoph Heiss <christoph@c8h4.io>

christoph-heiss committed:
sourcehut.buildsrht: 0.83.0 -> 0.86.10

Signed-off-by: Christoph Heiss <christoph@c8h4.io>

christoph-heiss committed:
sourcehut.gitsrht: 0.78.20 -> 0.84.2

Newer version of the gitsrht-api service call setrlimit() on startup,
thus allow it in the `SystemCallFilter` definition for the service.

Signed-off-by: Christoph Heiss <christoph@c8h4.io>

christoph-heiss committed:
sourcehut: fix `repos` path by using actual settings value

Signed-off-by: Christoph Heiss <christoph@c8h4.io>

christoph-heiss committed:
sourcehut: fix logging of git/hg ssh commands

Signed-off-by: Christoph Heiss <christoph@c8h4.io>

christoph-heiss committed:
sourcehut: remove `set -x` from ssh commands

Signed-off-by: Christoph Heiss <christoph@c8h4.io>

christoph-heiss committed:
sourcehut: explicitly disallow openssh to socket-active

This breaks the (already fragile) gitsrht-dispatch -> gitsrht-keys
command chain.

Signed-off-by: Christoph Heiss <christoph@c8h4.io>

christoph-heiss committed:
sourcehut: fix up some more bin paths

Signed-off-by: Christoph Heiss <christoph@c8h4.io>

nessdoor commented:
Thank you for taking the trouble of going through this much needed update.

I'm in the process of setting up my first Sourcehut forge, therefore I'm sorry if I won't be able to provide much knowledgeable feedback, but I'll try my best.

It seems that your PR suffers from the same problem as upstream where the `dispatch.sr.ht` service has been deprecated, but its presence in the `services` list triggers a deprecation assertion that is fatal to the build:
https://github.com/NixOS/nixpkgs/blob/818cfdd3c325e9fc725c6ce4dd1046978bbf8ebb/nixos/modules/services/misc/sourcehut/service.nix#L258-L259
Explicitly setting the `services` array is a viable workaround, but it is completely undocumented.

On this note, what about updating the manual with a working example as well, since the one that is written there is no longer valid? Maybe include something like the one in https://github.com/NixOS/nixpkgs/pull/133984. But maybe documentation can be updated as part of a separate PR, depending on how much work this one requires.

christoph-heiss commented:
> It seems that your PR suffers from the same problem as upstream where the `dispatch.sr.ht` service has been deprecated, but its presence in the `services` list triggers a deprecation assertion that is fatal to the build:
> [..]
> Explicitly setting the `services` array is a viable workaround, but it is completely undocumented.

Thanks for trying it out and catching that! In my config, I have set it explicitly, since I don't use/need all services. I'll fix it and will set set up a test server with all services enabled for the future.

> On this note, what about updating the manual with a working example as well, since the one that is written there is no longer valid?

Good point, I will do that too. Should not be too much of a hassle, I think.

nessdoor commented:
> I have set it explicitly, since I don't use/need all services.

Apparently, setting it explicitly was the old way of doing it, now replaced by the `<service>.enable` flag. I guess people never noticed the problem because everyone has legacy configs.

Another pain point is the need for all those OAuth IDs and secrets to be set manually, but I guess it would be non-trivial to come up with a safe way of generating them. Maybe this will have to be addressed in a future PR, but at least the documentation should be explicit about that.

christoph-heiss commented:
> Apparently, setting it explicitly was the old way of doing it, now replaced by the `<service>.enable` flag. I guess people never noticed the problem because everyone has legacy configs.

Well, then I'd suggest to explicitly deprecate setting the `services` array, so that people migrate. I will take a look at it, probably for the best in the long run anyway.

> Another pain point is the need for all those OAuth IDs and secrets to be set manually, but I guess it would be non-trivial to come up with a safe way of generating them. Maybe this will have to be addressed in a future PR, but at least the documentation should be explicit about that.

Yeah, secret management would definitely push this PR to far. That should be brainstormed separately, I'd rather have this PR go somewhere sooner than later.

christoph-heiss committed:
sourcehut: drop obsolete `services` array in favor of indivdual `enable` flags

Signed-off-by: Christoph Heiss <christoph@c8h4.io>

christoph-heiss committed:
sourcehut: disable IPv6 completely for tests

Signed-off-by: Christoph Heiss <christoph@c8h4.io>

nessdoor commented:
I can confirm that, at this moment, your PR seems to be working. I managed to bring up both `meta` and `git`, and they are reachable from the Internet.

I noticed you have chosen to deprecate `services` through an assertion that makes the build fail. Do you think it is better this way or do you feel like we could just be emitting a warning, so that people with old configs are not confronted with a build failure?

Apart from that, during the setup process I found two sources of frustration that, maybe, we could address in this PR.

The first one is that secret management is terribly inconsistent between secrets that are to be included in the INI files and those that, instead, are accessed by path. Those that are to be written to the INI files are the easiest to manage, as the activation script decrypts them as root and copies their contents verbatim to the INIs, meaning that the secret management utility that one uses (in my case, [agenix](https://github.com/ryantm/agenix)) doesn't have to treat them in any special way. The major issue with secrets that are accessed by path, though, is that _nowhere_ it is written that Sourcehut services run in heavily-restricted chroots, and therefore don't have access to the filesystem. If it weren't for your NixOS configs I found on your repo (thanks for that), I would have never guessed that I had to set the `serviceConfig.BindReadOnlyPaths` just to make the secrets accessible. And this passage is mandatory, as for some reason one cannot opt-out from the mailing service and needs to provide a GPG key pair (maybe another thing that can be improved?).

The other issue is that different services run under different users/groups, meaning that one has to either replicate the secrets and assign each one to the user of the service, or create an ad-hoc group with the various services' users, and assign it the ownership of the shared secrets.

Is there any way in which we could handle secrets better? Personally,  I went for an additional `srhtsec` group that owns the shared secrets, added `gitsrcht` and `metasrht` to it, then created a `/run/sourcehut/secrets` directory and instructed agenix to copy the GPG key there after decription.

Laslty, something still related to the GPG key but that can be solved through documentation: Sourcehut expects keys in armored format, not binary, but this is not said anywhere.

In the coming days, if I have time, I will try to spin up a `build` instance. I'll let you know how that goes. I haven't actually created any users, yet, so there may be hidden problems in there as well, but I don't expect any.

christoph-heiss commented:
> I noticed you have chosen to deprecate `services` through an assertion that makes the build fail. Do you think it is better this way or do you feel like we could just be emitting a warning, so that people with old configs are not confronted with a build failure?

Yeah, was my first thought too, but there isn't something like `mkDeprecatedOptionModule`, so I decided to simply drop that. But since one already need to set `enable = true` per individual service, users can simply drop that line from their config without having to change something else. So just breaking builds isn't _that_ bad either. 
I'll try to simply deprecate it instead of removing, even if IMHO has not a lot of value.


> The major issue with secrets that are accessed by path, though, is that _nowhere_ it is written that Sourcehut services run in heavily-restricted chroots, and therefore don't have access to the filesystem. If it weren't for your NixOS configs I found on your repo (thanks for that), I would have never guessed that I had to set the `serviceConfig.BindReadOnlyPaths` just to make the secrets accessible.

Thanks for pointing that out, I totally forgot about that issue due to having it "solved" in my config. I'll fix that :+1:

> Is there any way in which we could handle secrets better? Personally, I went for an additional `srhtsec` group that owns the shared secrets, added `gitsrcht` and `metasrht` to it, then created a `/run/sourcehut/secrets` directory and instructed agenix to copy the GPG key there after decription.

Hm .. having a separate group for all (shared) secrets does sounds like a good idea, I think. Maybe even something like `srhtservice` (or `_srhtservice`?), where all separate users like `gitsrht` etc. are members?


> Lastly, something still related to the GPG key but that can be solved through documentation: Sourcehut expects keys in armored format, not binary, but this is not told anywhere.

I will update the documentation w.r.t. that when I get around to that issue. The doc is definitely lacking lots of things. 

> In the coming days, if I have time, I will try to spin up a `build` instance. I'll let you know how that goes. I haven't actually created any users, yet, so there may be hidden problems in there as well, but I don't expect any.

Thanks! If something non-obvious comes up, I happily can help you out. Also, would be good to take notes of such things, so that I can improve it or add it to the documentation, as appropriate.

nessdoor commented:
> I'll try to simply deprecate it instead of removing, even if IMHO has not a lot of value.

I totally see your point, both solutions are acceptable, I think. I was just wondering which way would be better, but it's totally up to you.

> Hm .. having a separate group for all (shared) secrets does sounds like a good idea, I think. Maybe even something like srhtservice (or _srhtservice?), where all separate users like gitsrht etc. are members?

The group hack was the least convoluted solution to the problem without having to modify the module, but maybe it's not the cleanest as an upstream technique. I would not assume that all services share the same subset of secrets, therefore we would violate the least-privilege/need-to-know principle by doing that. Since we're building chroot jails for all services, we might as well copy and chown the necessary secrets inside each jail upon service activation, letting the derivation of each Sourcehut service choose which secrets it wants. I think this can be delegated to some logic near the generators for their Systemd services and INI configurations. Since the (encrypted) secrets are kept under the Nix store and tied to the system derivation, any change of secrets would entail a change of inputs, triggering a rebuild of the jails and reload of the services, so we shouldn't have any problems with stale secrets, either.

Well then, I'm off to setup this `build` thing. User management, repo creation and push/pull all work as expected. I am not going to test the public registration and mailing subsystem anytime soon, though.

fgaz commented:
> This issue ought to be resolved upstream, and I'll try to contact the NixOS image maintainer (@fgaz) to see if we can have an optional system attribute passed through the image.nix function.

Sure, feel free to send a patch to [sr.ht-dev](https://lists.sr.ht/~sircmpwn/sr.ht-dev) and cc me :)

nessdoor commented:
I was about to send you a private email, good timing!

Just let me learn how to send patches via email and I'll post it to the list.

nessdoor commented:
Time for another status report.

The modification to `image.nix` has been proposed upstream in [patch 43741](https://lists.sr.ht/~sircmpwn/sr.ht-dev/patches/43741). If it passes, we won't need to define the NixOS image ourselves.

As for the long stack trace generated while trying to submit build manifests, I might have found a related patch in [patches/redis-socket/core/0001-Fix-Unix-socket-support-in-RedisQueueCollector.patch](https://github.com/NixOS/nixpkgs/blob/0b6f3e17dda02c9cf1e8edc6e411ba37093f09e2/pkgs/applications/version-management/sourcehut/patches/redis-socket/core/0001-Fix-Unix-socket-support-in-RedisQueueCollector.patch) that is worth a look, I think. I will try to look into that either tomorrow or soon after.

Then, I decided to try out `paste`, which should be a rather simple service to setup and run. Turns out it doesn't work out of the box. The web UI is OK, but as soon as you try to create a paste, the following appears in the log:
```
gunicorn[24498]: [2023-08-19 17:27:36,708] ERROR in app: Exception on /new-paste [POST]
gunicorn[24498]: Traceback (most recent call last):
gunicorn[24498]:   File "/nix/store/gsx7dnmci82zr5m2ymh8z3mhzm64q5z4-python3-3.10.12-env/lib/python3.10/site-packages/urllib3/connection.py", line 174, in _new_conn
gunicorn[24498]:     conn = connection.create_connection(
gunicorn[24498]:   File "/nix/store/gsx7dnmci82zr5m2ymh8z3mhzm64q5z4-python3-3.10.12-env/lib/python3.10/site-packages/urllib3/util/connection.py", line 95, in create_connection
gunicorn[24498]:     raise err
gunicorn[24498]:   File "/nix/store/gsx7dnmci82zr5m2ymh8z3mhzm64q5z4-python3-3.10.12-env/lib/python3.10/site-packages/urllib3/util/connection.py", line 85, in create_connection
gunicorn[24498]:     sock.connect(sa)
gunicorn[24498]: ConnectionRefusedError: [Errno 111] Connection refused
gunicorn[24498]: During handling of the above exception, another exception occurred:
gunicorn[24498]: Traceback (most recent call last):
gunicorn[24498]:   File "/nix/store/gsx7dnmci82zr5m2ymh8z3mhzm64q5z4-python3-3.10.12-env/lib/python3.10/site-packages/urllib3/connectionpool.py", line 703, in urlopen
gunicorn[24498]:     httplib_response = self._make_request(
gunicorn[24498]:   File "/nix/store/gsx7dnmci82zr5m2ymh8z3mhzm64q5z4-python3-3.10.12-env/lib/python3.10/site-packages/urllib3/connectionpool.py", line 398, in _make_request
gunicorn[24498]:     conn.request(method, url, **httplib_request_kw)
gunicorn[24498]:   File "/nix/store/gsx7dnmci82zr5m2ymh8z3mhzm64q5z4-python3-3.10.12-env/lib/python3.10/site-packages/urllib3/connection.py", line 239, in request
gunicorn[24498]:     super(HTTPConnection, self).request(method, url, body=body, headers=headers)
gunicorn[24498]:   File "/nix/store/9c03r86hcdn43dm3hsgjirifvyzfkhwh-python3-3.10.12/lib/python3.10/http/client.py", line 1283, in request
gunicorn[24498]:     self._send_request(method, url, body, headers, encode_chunked)
gunicorn[24498]:   File "/nix/store/9c03r86hcdn43dm3hsgjirifvyzfkhwh-python3-3.10.12/lib/python3.10/http/client.py", line 1329, in _send_request
gunicorn[24498]:     self.endheaders(body, encode_chunked=encode_chunked)
gunicorn[24498]:   File "/nix/store/9c03r86hcdn43dm3hsgjirifvyzfkhwh-python3-3.10.12/lib/python3.10/http/client.py", line 1278, in endheaders
gunicorn[24498]:     self._send_output(message_body, encode_chunked=encode_chunked)
gunicorn[24498]:   File "/nix/store/9c03r86hcdn43dm3hsgjirifvyzfkhwh-python3-3.10.12/lib/python3.10/http/client.py", line 1038, in _send_output
gunicorn[24498]:     self.send(msg)
gunicorn[24498]:   File "/nix/store/9c03r86hcdn43dm3hsgjirifvyzfkhwh-python3-3.10.12/lib/python3.10/http/client.py", line 976, in send
gunicorn[24498]:     self.connect()
gunicorn[24498]:   File "/nix/store/gsx7dnmci82zr5m2ymh8z3mhzm64q5z4-python3-3.10.12-env/lib/python3.10/site-packages/urllib3/connection.py", line 205, in connect
gunicorn[24498]:     conn = self._new_conn()
gunicorn[24498]:   File "/nix/store/gsx7dnmci82zr5m2ymh8z3mhzm64q5z4-python3-3.10.12-env/lib/python3.10/site-packages/urllib3/connection.py", line 186, in _new_conn
gunicorn[24498]:     raise NewConnectionError(
gunicorn[24498]: urllib3.exceptions.NewConnectionError: <urllib3.connection.HTTPConnection object at 0x7fa75f184430>: Failed to establish a new connection: [Errno 111] Connection refused
gunicorn[24498]: During handling of the above exception, another exception occurred:
gunicorn[24498]: Traceback (most recent call last):
gunicorn[24498]:   File "/nix/store/gsx7dnmci82zr5m2ymh8z3mhzm64q5z4-python3-3.10.12-env/lib/python3.10/site-packages/requests/adapters.py", line 486, in send
gunicorn[24498]:     resp = conn.urlopen(
gunicorn[24498]:   File "/nix/store/gsx7dnmci82zr5m2ymh8z3mhzm64q5z4-python3-3.10.12-env/lib/python3.10/site-packages/urllib3/connectionpool.py", line 787, in urlopen
gunicorn[24498]:     retries = retries.increment(
gunicorn[24498]:   File "/nix/store/gsx7dnmci82zr5m2ymh8z3mhzm64q5z4-python3-3.10.12-env/lib/python3.10/site-packages/urllib3/util/retry.py", line 592, in increment
gunicorn[24498]:     raise MaxRetryError(_pool, url, error or ResponseError(cause))
gunicorn[24498]: urllib3.exceptions.MaxRetryError: HTTPConnectionPool(host='localhost', port=5111): Max retries exceeded with url: /query (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7fa75f184430>: Failed to establish a new connection: [Errno 111] Connection refused'))
gunicorn[24498]: During handling of the above exception, another exception occurred:
gunicorn[24498]: Traceback (most recent call last):
gunicorn[24498]:   File "/nix/store/gsx7dnmci82zr5m2ymh8z3mhzm64q5z4-python3-3.10.12-env/lib/python3.10/site-packages/flask/app.py", line 2529, in wsgi_app
gunicorn[24498]:     response = self.full_dispatch_request()
gunicorn[24498]:   File "/nix/store/gsx7dnmci82zr5m2ymh8z3mhzm64q5z4-python3-3.10.12-env/lib/python3.10/site-packages/flask/app.py", line 1825, in full_dispatch_request
gunicorn[24498]:     rv = self.handle_user_exception(e)
gunicorn[24498]:   File "/nix/store/gsx7dnmci82zr5m2ymh8z3mhzm64q5z4-python3-3.10.12-env/lib/python3.10/site-packages/flask/app.py", line 1823, in full_dispatch_request
gunicorn[24498]:     rv = self.dispatch_request()
gunicorn[24498]:   File "/nix/store/gsx7dnmci82zr5m2ymh8z3mhzm64q5z4-python3-3.10.12-env/lib/python3.10/site-packages/flask/app.py", line 1799, in dispatch_request
gunicorn[24498]:     return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)
gunicorn[24498]:   File "/nix/store/gsx7dnmci82zr5m2ymh8z3mhzm64q5z4-python3-3.10.12-env/lib/python3.10/site-packages/srht/oauth/__init__.py", line 43, in wrapper
gunicorn[24498]:     return f(*args, **kwargs)
gunicorn[24498]:   File "/nix/store/gsx7dnmci82zr5m2ymh8z3mhzm64q5z4-python3-3.10.12-env/lib/python3.10/site-packages/pastesrht/blueprints/public.py", line 96, in new_paste_POST
gunicorn[24498]:     paste_id = create_paste(valid, files, visibility)
gunicorn[24498]:   File "/nix/store/gsx7dnmci82zr5m2ymh8z3mhzm64q5z4-python3-3.10.12-env/lib/python3.10/site-packages/pastesrht/blueprints/public.py", line 45, in create_paste
gunicorn[24498]:     resp = op.execute("paste.sr.ht", valid=valid)
gunicorn[24498]:   File "/nix/store/gsx7dnmci82zr5m2ymh8z3mhzm64q5z4-python3-3.10.12-env/lib/python3.10/site-packages/srht/graphql/client.py", line 79, in execute
gunicorn[24498]:     r = requests.post(f"{origin}/query",
gunicorn[24498]:   File "/nix/store/gsx7dnmci82zr5m2ymh8z3mhzm64q5z4-python3-3.10.12-env/lib/python3.10/site-packages/requests/api.py", line 115, in post
gunicorn[24498]:     return request("post", url, data=data, json=json, **kwargs)
gunicorn[24498]:   File "/nix/store/gsx7dnmci82zr5m2ymh8z3mhzm64q5z4-python3-3.10.12-env/lib/python3.10/site-packages/requests/api.py", line 59, in request
gunicorn[24498]:     return session.request(method=method, url=url, **kwargs)
gunicorn[24498]:   File "/nix/store/gsx7dnmci82zr5m2ymh8z3mhzm64q5z4-python3-3.10.12-env/lib/python3.10/site-packages/requests/sessions.py", line 589, in request
gunicorn[24498]:     resp = self.send(prep, **send_kwargs)
gunicorn[24498]:   File "/nix/store/gsx7dnmci82zr5m2ymh8z3mhzm64q5z4-python3-3.10.12-env/lib/python3.10/site-packages/requests/sessions.py", line 703, in send
gunicorn[24498]:     r = adapter.send(request, **kwargs)
gunicorn[24498]:   File "/nix/store/gsx7dnmci82zr5m2ymh8z3mhzm64q5z4-python3-3.10.12-env/lib/python3.10/site-packages/requests/adapters.py", line 519, in send
gunicorn[24498]:     raise ConnectionError(e, request=request)
gunicorn[24498]: requests.exceptions.ConnectionError: HTTPConnectionPool(host='localhost', port=5111): Max retries exceeded with url: /query (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7fa75f184430>: Failed to establish a new connection: [Errno 111] Connection refused'))
```
The cause behind this failure is that `paste` tries to contact its API backend, which, for some reason, doesn't exist. Apparently, `pastesrht` has been packaged incorrectly, as the `buildGoModule` for its API component is absent, resulting in a half-baked package. The fix is trivial, as the other services give plenty of examples on how to compile these hybrid Python/Go packages correctly.

This defect could have been caught by a simple test waiting for a socket to appear on the API port, so I suggest we add something along those lines to verify that these backend components are actually up and running. I guess very few people self-host their pastebins on NixOS machines...

tomberek commented:
Getting all the services to work together and configured was a hurdle last time I was looking at this. There were some cases where there was so much service isolation that they couldn't communicate. Are there tests we can add to to make this less fragile?

nessdoor commented:
I would advise some narrow-scoped integration tests that carry out some basic inter-component transaction, in addition to simple tests to verify that all components are up.

For example, pushing a new repo to `git` would test the full integration between `git`, `meta` and any intervening databases (you know, that thing that goes under the name of "thread testing"), and that would suffice, I think. After all, upstream already carries out their set of application tests, so we just need to ascertain that our inter-service connections work.

By the way, @tomberek, since you are the original maintainer of this module, do you have any working configuration for a multi-service Sourcehut installation to show us and that can guide us through some of the basic pitfalls? Many of the hurdles that I am encountering are being solved by reading other people's configs.

nessdoor commented:
I don't know if I am stating the obvious, but it seems that the `builds` failure is due to the fact that a Redis endpoint URL meant for Celery is leaking to redis-go, which should have its own URL, instead.
At the moment, I don't have enough disk space to regenerate a `builds` installation, I'll try to confirm my suspects as soon as I repartition my server and can generate a `config.ini`. `default.nix` and `service.nix` are too freaking confusing for me to inspect manually, I already spent some hours on those and didn't spot anything suspicious; too many recursive merges, options defined here and there, attribute overrides all over the place and closures being passed around.

Also, one day we should refactor this module to respect RFC42, because the amount of options is hideous and the repetition is disturbing.

nessdoor commented:
Found it.

At [line 22 of `api/graph/celery.go`](https://git.sr.ht/~sircmpwn/builds.sr.ht/tree/b456338259c82664eb0e2ea1b54fd8afe97fa49f/item/api/graph/celery.go#L22), `builds` pulls the `redis` config option from the INI file, which is the URI for Celery workers.

When running in the default local configuration, the Redis store is on the same machine as the worker, so the connection is established via socket, resulting in a URI with a `redis+socket://` schema. The Celery library for Python mandates this schema (although it is currently undocumented, see https://github.com/celery/celery/pull/6988), but unfortunately that is not the case for the Go library.
More specifically, the deprecated functions [`NewRedisCeleryBroker`](https://github.com/gocelery/gocelery/blob/825d89059344006c104412e628d456b3a111a6a5/redis_broker.go#L33) and [`NewRedisCeleryBackend`](https://github.com/gocelery/gocelery/blob/825d89059344006c104412e628d456b3a111a6a5/redis_backend.go#L31) of [gocelery](https://github.com/gocelery/gocelery) ultimately rely on a call to the deprecated [`NewRedisPool`](https://github.com/gocelery/gocelery/blob/825d89059344006c104412e628d456b3a111a6a5/redis_broker.go#L90C24-L90C24), which in turn tries to open a connection to Redis by passing the connection URI to the [`DialURL`](https://github.com/gomodule/redigo/blob/d6854479365f0307560fa28e18e2bd0634b05229/redis/conn.go#L320) function of [redigo](https://github.com/gomodule/redigo).
The `invalid redis URL scheme: redis+socket` error message is the result of the `DialURLContext` function of redigo [not knowing what to do with that schema](https://github.com/gomodule/redigo/blob/d6854479365f0307560fa28e18e2bd0634b05229/redis/conn.go#L335).

The reason why nobody reported this bug before is probably because everyone using `builds` is either connecting to the local Redis instance via loopback, using a standard `redis://` schema, or is running a remote Redis store.

What can we do now? I think that, for the time being, we should either patch the Go source code with something that mangles the schema into the right form (difficult), or we should make it so that the local Redis instance, at least for `builds`, is accessed through a loopback connection, so that both Python Celery and redigo are happy (easy).
We should also think about informing upstream of this issue, since it is something they should either explicitly state as unsupported in their documentation, or manage in their code. Maybe the problem will disappear once they refactor to remove the deprecated functions, as they will have to interface with redigo directly, but, given the fact that this bug has gone unnoticed for so long, I wouldn't count on that.

nessdoor commented:
With a simple modification to my system configuration, I was able to make `builds` communicate with its Redis instance via loopback, and can confirm that the problem is solved and the web UI behaves correctly, with the build being submitted.

Sadly, it seems that `builds` suffers from a defect in its derivation. Due to the commenting of [line 948 of `default.nix`](https://github.com/NixOS/nixpkgs/blob/0b6f3e17dda02c9cf1e8edc6e411ba37093f09e2/nixos/modules/services/misc/sourcehut/default.nix#L948) at the end of year 2021, the [`control`](https://git.sr.ht/~sircmpwn/builds.sr.ht/tree/b456338259c82664eb0e2ea1b54fd8afe97fa49f/item/images/control) script in charge of running the images is not getting copied to the `lib/images` directory of the service derivation, resulting in a
```
run_build panic: boot: fork/exec /nix/store/<hash>-buildsrht-worker-images/images/control: no such file or directory
```
In addition, the `control` script contains many hard-coded absolute paths, meaning that it hasn't been fixpath'ed nor parameterized, making it broken by default.

Again, this is an easy issue to solve, as the reason for commenting that line is no longer applicable, any future re-occurrence of the problem can be easily side-stepped, and the absolute paths can all be replaced by store paths and build-time parameters. What troubles me the most is that, once again, this bug has been sitting there for more than a year, with 3 intervening NixOS releases shipping broken code and no failure detected by the testing suite.

That said, I begin to wonder if it is meaningful to hold back on the merge of this PR, @christoph-heiss. Your updates are not breaking any of the core services that people seem to actually use (`git` and `meta`), and the other services seem to be broken from the start, so a version bump wouldn't hurt them, either. I think that, strategically speaking, it would make sense to have this global minor version bump committed, and then flesh-out the rest in subsequent PRs (after all, we are on `unstable`). Otherwise, I feel like the testing process of this PR would take too long and make it disappear in the PR soup from hell, wasting everybody's time.

Or, maybe we should ask for help on Discourse, and have other Sourcehut administrators provide suggestions.

@fgaz, sorry to trouble you again, but I noticed that you lamented the lack of meaningful Minio integration in https://github.com/NixOS/nixpkgs/issues/164440. Does this mean you are using any other service besides `meta` and `git` that are not Git LFS? Can you provide some insight into any other broken things not caught by mainstream users?

fgaz commented:
> @fgaz, sorry to trouble you again, but I noticed that you lamented the lack of meaningful Minio integration in #164440. Does this mean you are using any other service besides `meta` and `git` that are not Git LFS? Can you provide some insight into any other broken things not caught by mainstream users?

iirc I stumbled upon that issue while attempting to test a pages.sr.ht patch on which I gave up, so I don't have anything else sorry

nessdoor commented:
It's fine it's fine, better for you. Thanks!

nessdoor commented:
Success reports follow.

First of all, the fix of the `paste` issue was trivial and, after building the API server and setting its unit file up, everything works fine.

The patch to the `image.nix` expression for building NixOS builder images has been revised and re-proposed upstream as patch number [44014](https://lists.sr.ht/~sircmpwn/sr.ht-dev/patches/44014). Awaiting further feedback or application.

As for `builds` itself, after having:
- massaged the `/images` directory into better shape
- wrapped the `control` script
- exposed `/bin` to the chrooted worker
- exposed a Docker socket to such worker
- patched `control` to follow symlinks when mounting volumes

I finally succeeded in running the NixOS image. Great success!

Still a problem remains for `build`, though: for some reason, the API daemon is incapable of retrieving logs. All its requests to `/query/log/<build-id>/log` are met with a 502 response. I suspect a simple misconfiguration of Nginx, but I am not familiar with webservers, so it may take some time do debug.

All (temporary) fixes that I applied can be found at [my Nixpkgs fork](https://git.entropic.network/~nessdoor/nixpkgs/tree/pkgs/sourcehut).

I have yet to decide which service to test next, but probably `todo`, as it would be a nice companion for my branch of temporary Sourcehut-on-NixOS patchwork.

I will write a checklist with all we've tested and with all we are yet to test, so that people may come over and help out without going through my walls of text.

nessdoor commented:
Oh, and I just noticed https://github.com/NixOS/nixpkgs/pull/202608. Sad story indeed, I can resonate with much of that frustration.

@apfelkuchen6 I would love to hear your opinion! If you have time and are not nauseated by this module, of course...

nessdoor commented:
- [ ] Integrations
  - [x] PostgreSQL
  - [x] Redis
  - [ ] Postfix
  - [ ] Minio (see https://github.com/NixOS/nixpkgs/issues/164440)
- [x] `meta`
- [ ] `builds`
  - [x] VM boot succeeds
  - [x] VM functional (compute + network)
  - [ ] Logs visible on WebUI
- [x] `git`
- [ ] `hg`
- [ ] `hub`
- [ ] `lists`
- [ ] `man`
- [ ] `pages`
- [x] `paste`
- [ ] `todo`

tomberek commented:
I know this is a large amount of work. I apologize for not testing and reviewing yet.

nessdoor commented:
Report: `hub` is working without need for additional modifications.

Although it seems it has a backend component that is not activated, I really don't know what purpose it serves, as everything seems to be working without it. Maybe issues will crop up while integrating other services like `todo` and `lists`, but for now I will give it a passing mark.

@tomberek everyone has better things to do than reviewing NixOS modules, so it's understandable, don't worry!
In any case, when you will have time to get your hands on this, please take into account that, even if some of the updated services won't work right off the bat, chances are that they didn't work at all in the first place, so their malfunction should not be a blocker for this PR. In addition, I have a bunch of fixes ready for at least half of them, so as soon as this PR lands, I can rapid-fire fixup PRs and stabilize much of the Sourcehut module in a short time. Maybe we could begin moving in that direction by undrafting this PR?

christoph-heiss commented:
First of, thank you **all** for helping with this PR and showing that these modules are still very much used!

Sorry for my quietness these last weeks due to vacation.

> That said, I begin to wonder if it is meaningful to hold back on the merge of this PR, @christoph-heiss. Your updates are not breaking any of the core services that people seem to actually use (git and meta), and the other services seem to be broken from the start, so a version bump wouldn't hurt them, either. I think that, strategically speaking, it would make sense to have this global minor version bump committed, and then flesh-out the rest in subsequent PRs (after all, we are on unstable). Otherwise, I feel like the testing process of this PR would take too long and make it disappear in the PR soup from hell, wasting everybody's time.

> even if some of the updated services won't work right off the bat, chances are that they didn't work at all in the first place, so their malfunction should not be a blocker for this PR. In addition, I have a bunch of fixes ready for at least half of them, so as soon as this PR lands, I can rapid-fire fixup PRs and stabilize much of the Sourcehut module in a short time. Maybe we could begin moving in that direction by undrafting this PR?

Later on today I will rebase this and re-update all of these components and then un-draft this! So hopefully it can finally get merged, such that we can pile all the other fixes on top.

nessdoor commented:
I can confirm that all the fixes I have accumulated keep working after the rebase.

Just a minor thingy: I use `nixpkgs-fmt` to format Nix code for Nixpkgs automatically, and it seems that neither `default.nix` nor `service.nix` follow the formatting standard enforced by the tool. Do you mind trying to reformat those files with the same tool? But this is just optional, don't stress too much about it.

christoph-heiss commented:
> Just a minor thingy: I use `nixpkgs-fmt` to format Nix code for Nixpkgs automatically, and it seems that neither `default.nix` nor `service.nix` follow the formatting standard enforced by the tool. Do you mind trying to reformat those files with the same tool? But this is just optional, don't stress too much about it.

Can happily do that. Both weren't formatted before, so this results in a rather big churn, as most parts of these files get re-written. So I think as a separate PR would be better?
I can immediately send one as soon as this gets merged.

christoph-heiss commented:
Ping. It's been 1.5 weeks, and as sooner this get's merged, the sooner we can all send further PRs improving. There's a lot in the pipeline by several people regarding that :)

nessdoor commented:
@christoph-heiss try drawing attention over at Discourse

h7x4 commented on nixos/modules/services/misc/sourcehut/default.nix:
```diff
@@ -1066,10 +1064,11 @@ in
+      extraServices.gitsrht-api.serviceConfig = {
+        Restart = "always";
+        RestartSec = "5s";
+        ExecStart = "${pkgs.sourcehut.gitsrht}/bin/gitsrht-api -b ${cfg.listenAddress}:${toString (cfg.git.port + 100)}";
```
```suggestion
        ExecStart = "${pkgs.sourcehut.gitsrht}/bin/gitsrht-api -b ${cfg.settings."git.sr.ht".api-origin}";
```

h7x4 commented on nixos/modules/services/misc/sourcehut/default.nix:
```diff
@@ -77,6 +77,14 @@ let
       apply = s: "<" + toString s;
     };
+    api-origin = mkOption {
+      description = lib.mdDoc "Origin URL for API, 100 more than web.";
```
I know you just moved this, but could we change this description? It doesn't inherently make sense. Anyone looking at this description will also be able to see the default, so I propose we just remove the last part.

```suggestion
      description = lib.mdDoc "Origin URL for the API";
```

h7x4 commented on nixos/modules/services/misc/sourcehut/default.nix:
```diff
@@ -792,15 +786,28 @@ in
+            "chown ${cfg.hg.user}:${cfg.hg.group} /var/log/sourcehut/hgsrht-keys"
+          ]);
         serviceConfig = {
+          LogsDirectory = mkIf (cfg.git.enable || cfg.hg.enable) "sourcehut";
```
I'm inclined to say we should just create this directory unconditionally. It's less maintenance in case more things start spitting logs, in theory a miniscule eval time improvement, and having an empty log dir shouldn't hurt anyone.

h7x4 commented on nixos/modules/services/misc/sourcehut/service.nix:
```diff
@@ -108,7 +108,7 @@ let
       SystemCallFilter = [
         "@system-service"
-        "~@aio" "~@keyring" "~@memlock" "~@privileged" "~@resources" "~@timer"
+        "~@aio" "~@keyring" "~@memlock" "~@privileged" "~@timer"
```
I believe this is a shorter way to say the same thing.

```suggestion
        "~@aio @keyring @memlock @privileged @timer"
```

christoph-heiss replied:
This can be done in a separate PR. I'm not willing to introduce _even more_ invasive changes here at this point, just to be sure.

h7x4 commented on nixos/modules/services/misc/sourcehut/default.nix:
```diff
@@ -77,6 +77,14 @@ let
+      description = lib.mdDoc "Origin URL for API, 100 more than web.";
+      type = types.str;
+      default = "http://${cfg.listenAddress}:${toString (cfg.${srv}.port + 100)}";
+      defaultText = lib.literalMD ''
+        `"http://''${`[](#opt-services.sourcehut.listenAddress)`}:''${toString (`[](#opt-services.sourcehut.${srv}.port)` + 100)}"`
+      '';
+    };
```
Let's define `opt = options.services.sourcehut;` alongside `cfg`. Not only is it a bit shorter - it also ensures the option actually exists.

```suggestion
      defaultText = lib.literalExpression ''
        "http://''${config.${opt.listenAddress}}:''${toString (config.${opt.${srv}.port} + 100)}"
      '';
    };
```

christoph-heiss replied:
Again, I'd rather do this as a separate PR. It's not breaking anything currently, so let's keep at this for this PR.

h7x4 reviewed (changes requested):
I have not looked at the packages, only a few comments on the module stuff.

It also seems to have a few merge conflicts going on.

christoph-heiss commented:
> I have not looked at the packages, only a few comments on the module stuff.

Thanks for the review in any case!
I will go through the comments over the weekend and address them.  

> It also seems to have a few merge conflicts going on.

Yeah, I'll also resolve this. This PR is now open for 3 months, so that stuff is bound to happen ..

tomberek commented:
I'm doing a build and testing. I've not been using the module recently, so I may not be the best person to assess the impact of the change on users of the module, other than basic review.

Is there any reason not to merge if we are passing the tests? Do we need any release notes or doc changes?

nessdoor commented:
The branch-off for the new release is just 9 days away, it would be really nice to have these changes incorporated before then. 

@christoph-heiss any news?

christoph-heiss committed:
sourcehut: reword `api-origin` option description

Signed-off-by: Christoph Heiss <christoph@c8h4.io>

christoph-heiss committed:
sourcehut: create logs directory unconditionally

An empty log directory, in case it stays unused, does not hurt anyone.

Signed-off-by: Christoph Heiss <christoph@c8h4.io>

christoph-heiss committed:
sourcehut: add overrides for `flask-sqlalchemy` and `factory-boy`

These are needed, as the used sourcehut version is not compatible with
the newer major-releases for both packages.

Signed-off-by: Christoph Heiss <christoph@c8h4.io>

christoph-heiss committed:
sourcehut: use systemd.tmpfiles instead of manually creating logfiles

Signed-off-by: Christoph Heiss <christoph@c8h4.io>

christoph-heiss committed:
sourcehut: de-duplicate nginx `add_header` directives

Signed-off-by: Christoph Heiss <christoph@c8h4.io>

christoph-heiss committed:
sourcehut: fix postgresql database permission for postgresql >= 15

Signed-off-by: Christoph Heiss <christoph@c8h4.io>

christoph-heiss commented:
I have rebased this branch on the current master and addressed most of @h7x4's feedback.
A full rebuild and all tests pass cleanly.

I also had to add overrides for two more packages, which got major releases in the meantime. And a fix for PostgreSQL >= 15 was also needed.

This PR has now gone **way beyond** its scope. I'm not willing to do any more changes, as otherwise this will stay open forever. Either it gets merged as-is or never.

I cannot justify wasting more time & energy, just to constantly keep up with the master branch.

sarcasticadmin commented on nixos/modules/services/misc/sourcehut/default.nix:
```diff
@@ -1252,55 +1251,30 @@ in
+        ];
+        environment.systemPackages = optional cfg.meta.enable
+          (pkgs.writeShellScriptBin "metasrht-manageuser" ''
+            set -eux
```
another one `set -x` that could potentially be dropped

sarcasticadmin commented on nixos/modules/services/misc/sourcehut/service.nix:
```diff
@@ -363,6 +377,20 @@ in
         extraService
       ])) extraServices)
+
+      # Work around 'pq: permission denied for schema public' with postgres v15, until a
```
thanks for the note here

sarcasticadmin reviewed (commented):
~I just a couple of small observations around other places that still had `set -x`~

> This PR has now gone way beyond its scope. I'm not willing to do any more changes, as otherwise this will stay open forever. Either it gets merged as-is or never.

Just saw this comment. Please ignore since I totally can understand where your coming from

tomberek commented:
Result of `nixpkgs-review pr 245394` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
    <li>tests.nixos-functions.nixos-test</li>
  </ul>
</details>
<details>
  <summary>21 packages built:</summary>
  <ul>
    <li>sourcehut.buildsrht</li>
    <li>sourcehut.buildsrht.dist</li>
    <li>sourcehut.coresrht</li>
    <li>sourcehut.coresrht.dist</li>
    <li>sourcehut.gitsrht</li>
    <li>sourcehut.gitsrht.dist</li>
    <li>sourcehut.hgsrht</li>
    <li>sourcehut.hgsrht.dist</li>
    <li>sourcehut.hubsrht</li>
    <li>sourcehut.hubsrht.dist</li>
    <li>sourcehut.listssrht</li>
    <li>sourcehut.listssrht.dist</li>
    <li>sourcehut.mansrht</li>
    <li>sourcehut.mansrht.dist</li>
    <li>sourcehut.metasrht</li>
    <li>sourcehut.metasrht.dist</li>
    <li>sourcehut.pagessrht</li>
    <li>sourcehut.pastesrht</li>
    <li>sourcehut.pastesrht.dist</li>
    <li>sourcehut.todosrht</li>
    <li>sourcehut.todosrht.dist</li>
  </ul>
</details>

tomberek reviewed (approved):
Build and functions as expected. NixOS tests pass.
(also did testing on x86_64-linux)

tomberek commented:
@christoph-heiss : thanks so much for the work. The constant update and rebase is not fun. (this is mostly my fault, i've not been using or testing along the way, making it much slower to do review).

nessdoor commented:
I'll try to make it quick, but I won't assure that I will be able to land all the fixes before the branch-off. Great thing that at least we have a refreshed package set and fixed service module, now!

Thank you, see you soon!

nessdoor commented:
Oh right, can someone make sure that #201424 and #201425 are indeed no longer applicable? I can say that #201425 is probably solved, by now, but I have little clue about the other one.

christoph-heiss commented:
> Oh right, can someone make sure that #201424 and #201425 are indeed no longer applicable? I can say that #201425 is probably solved, by now, but I have little clue about the other one.

Yes, #201424 and #201425 indeed have been fixed this PR.
